### DISCUSSION TITLE
jvm/gc-collector

### DISCUSSION THREAD

**Discussion by giscus** [itwanger/tobebetterjavaer-giscus#281] [Announcements]
# jvm/gc-collector

本篇内容我们主要介绍了 CMS、G1 和 ZGC 三种垃圾收集器，它们都是分区收集器，都是为了降低 GC 停顿时间而生的，但是它们各有优缺点，我们可以根据业务场景选择合适的垃圾收集器。

https://javabetter.cn/jvm/gc-collector.html

**Comment by qqq1521902442** [comment]
G1回收器的部分，前面讲进行老年代的垃圾回收时，G1 使用标记-整理算法。这个过程分为两个阶段：标记存活的对象和整理（压缩）堆空间。通过整理，G1 能够避免内存碎片化，提高内存利用率。
年轻代的垃圾回收（Minor GC）使用复制算法。
后面为啥又讲G1是基于标记-清除算法的，因此需要STW呢？

**Reply by chaofengdev** [reply]
是的。我没看懂这篇和上篇文章的关系。

**Reply by chaofengdev** [reply]
所以我的理解是，上篇文章是垃圾收集的简单模型的科普？

**Reply by chaofengdev** [reply]
我大概理解了，上篇文章讲的是垃圾收集的常见方法，而这篇文章讲的是具体的垃圾收集器。具体的垃圾收集器比较复杂。

**Comment by chaofengdev** [comment]
我比较费解的是：`G1 中存在三种 GC 模式，分别是 Young GC、Mixed GC 和 Full GC。`
这些和上篇`深入理解垃圾回收机制`里提到的minor jc和major jc有什么关系吗？

**Comment by Yolo-XYX** [comment]
讲的不尽人意，尤其zgc那一部分，很多错误以及参考别人直接搬过来上下没有逻辑

**Comment by VirtualBool** [comment]
cms使用了复制算法吗？应该没有吧

**Comment by ExcuMe** [comment]
G1部分感觉没讲透啊，比如为什么需要分成多个Region、为什么需要Mixed GC这一步

**Reply by ExcuMe** [reply]
还有具体的回收过程时怎样的
